File: adapters/express/swapbook.js
"use strict";
// Swapbook adapter for Express.
//
// Exposes the Swapbook protocol (manifest / preview / mocks / mock) as an
// Express router. Render-agnostic: a variant's render is (args) => htmlString,
// so it works with any template engine or raw strings. Mount it:
//   app.use(registry.router());
const express = require("express");

function slug(s) {
  return s.toLowerCase().replace(/[^a-z0-9]+/g, "-").replace(/-+/g, "-").replace(/^-|-$/g, "");
}

function splitRoute(route) {
  const i = route.indexOf(" ");
  return i === -1 ? ["GET", route] : [route.slice(0, i), route.slice(i + 1)];
}

function control(name, type = "text", def = null, options) {
  const c = { name, type, default: def };
  if (options) c.options = options;
  return c;
}

// Pass a non-2xx status (422, 500, ...) to preview a component's error state.
function mock(route, render, status = 200) {
  const [verb, path] = splitRoute(route);
  return { verb: verb.toUpperCase(), path, render, status };
}

// A named preview width, added to the built-in full/tablet/phone.
function viewport(name, w) {
  return { name, w };
}

// Play steps: scripted interactions + assertions run against the preview.
function click(target) { return { action: "click", target }; }
function type(target, value) { return { action: "type", target, value }; }
function expectText(target, text) { return { action: "expect-text", target, text }; }
function expectVisible(target) { return { action: "expect-visible", target }; }
function wait(target) { return { action: "wait", target }; }

function variant(name, render, opts = {}) {
  return {
    name, render,
    controls: opts.controls || [],
    docs: opts.docs || "",
    mocks: opts.mocks || [],
    play: opts.play || [],
  };
}

function coerce(controls, query) {
  const args = {};
  for (const c of controls) {
    const raw = query[c.name];
    if (raw === undefined) { args[c.name] = c.default; continue; } // absent -> default
    if (c.type === "number") {
      const n = parseFloat(raw);
      args[c.name] = Number.isNaN(n) ? c.default : n;
    } else if (c.type === "bool") {
      args[c.name] = raw === "true" || raw === "1" || raw === "on";
    } else {
      args[c.name] = raw;
    }
  }
  return args;
}

class Registry {
  constructor(opts = {}) {
    this.htmxSrc = opts.htmxSrc || "";
    this.cssSrc = opts.cssSrc || "";
    this.jsSrc = opts.jsSrc || "";
    this.viewports = opts.viewports || [];
    this._stories = [];
    this._globalMocks = [];
  }

  register(name, variants, opts = {}) {
    this._stories.push({ id: slug(name), name, group: opts.group || "", docs: opts.docs || "", variants });
    return this;
  }

  // Registry-level mock merged into every variant; a variant's own mock for the
  // same route wins.
  mock(route, render, status = 200) {
    this._globalMocks.push(mock(route, render, status));
    return this;
  }

  _mocksFor(v) { return this._globalMocks.concat(v.mocks); }
  _find(sid, vname) {
    const s = this._stories.find((x) => x.id === sid);
    return s && s.variants.find((v) => v.name === vname);
  }

  _manifest() {
    const vmeta = (v) => {
      const m = { name: v.name, controls: v.controls, docs: v.docs };
      if (v.play && v.play.length) m.play = v.play;
      return m;
    };
    const out = {
      htmxSrc: this.htmxSrc, cssSrc: this.cssSrc, jsSrc: this.jsSrc,
      stories: this._stories.map((s) => ({
        id: s.id, name: s.name, group: s.group, docs: s.docs,
        variants: s.variants.map(vmeta),
      })),
    };
    if (this.viewports.length) out.viewports = this.viewports;
    return out;
  }

  router() {
    const r = express.Router();
    r.get("/_swapbook/manifest.json", (req, res) => res.json(this._manifest()));
    r.get("/_swapbook/preview/:sid/:vname", (req, res) => {
      const v = this._find(req.params.sid, req.params.vname);
      if (!v) return res.sendStatus(404);
      res.type("html").send(v.render(coerce(v.controls, req.query)));
    });
    r.get("/_swapbook/mocks/:sid/:vname", (req, res) => {
      const v = this._find(req.params.sid, req.params.vname);
      if (!v) return res.sendStatus(404);
      res.json(this._mocksFor(v).map((m, i) => ({ verb: m.verb, path: m.path, index: i })));
    });
    r.all("/_swapbook/mock/:sid/:vname/:index", (req, res) => {
      const v = this._find(req.params.sid, req.params.vname);
      if (!v) return res.sendStatus(404);
      const mk = this._mocksFor(v)[parseInt(req.params.index, 10)];
      if (!mk) return res.sendStatus(404);
      res.status(mk.status || 200).type("html").send(mk.render({}));
    });
    return r;
  }
}

module.exports = { Registry, control, mock, variant, viewport, click, type, expectText, expectVisible, wait };

File: adapters/flask/swapbook.py
"""Swapbook adapter for Flask.

Exposes the Swapbook protocol (manifest / preview / mocks / mock) as a Flask
blueprint. Render-agnostic: a variant is a callable ``(args) -> str`` returning
HTML, so it works with Jinja templates or raw strings. Mount it:

    app.register_blueprint(registry.blueprint)
"""
import re

from flask import Blueprint, Response, jsonify, request

Args = dict  # {control_name: coerced_value}


def slug(s: str) -> str:
    return re.sub(r"-+", "-", re.sub(r"[^a-z0-9]+", "-", s.lower())).strip("-")


def control(name, type="text", default=None, options=None):
    """Declare an editable control (knob). type: text|number|bool|select."""
    c = {"name": name, "type": type, "default": default}
    if options:
        c["options"] = options
    return c


def mock(route, render, status=200):
    """A canned response: route is "VERB /path", render is (args)->str. Pass a
    non-2xx status (422, 500, ...) to preview a component's error state."""
    verb, _, p = route.partition(" ") if " " in route else ("GET", "", route)
    return {"verb": verb.upper(), "path": p or route, "render": render, "status": status}


def viewport(name, w):
    """A named preview width, added to the built-in full/tablet/phone."""
    return {"name": name, "w": w}


# Play steps: scripted interactions + assertions run against the preview.
def click(target):
    return {"action": "click", "target": target}


def type_(target, value):
    return {"action": "type", "target": target, "value": value}


def expect_text(target, text):
    return {"action": "expect-text", "target": target, "text": text}


def expect_visible(target):
    return {"action": "expect-visible", "target": target}


def wait(target):
    return {"action": "wait", "target": target}


def variant(name, render, controls=None, docs="", mocks=None, play=None):
    """One rendered state. render is a callable (args: dict) -> html str."""
    return {"name": name, "render": render, "controls": controls or [], "docs": docs, "mocks": mocks or [], "play": play or []}


def _coerce(controls, args_get):
    args = {}
    for c in controls:
        name = c["name"]
        raw = args_get(name)
        if raw is None:  # absent -> default; present-but-empty is a real value
            args[name] = c.get("default")
            continue
        t = c.get("type")
        if t == "number":
            try:
                args[name] = float(raw)
            except ValueError:
                args[name] = c.get("default")
        elif t == "bool":
            args[name] = raw in ("true", "1", "on")
        else:
            args[name] = raw
    return args


class Registry:
    def __init__(self, htmx_src="", css_src="", js_src="", viewports=None):
        self.htmx_src, self.css_src, self.js_src = htmx_src, css_src, js_src
        self.viewports = viewports or []
        self._stories = []
        self._global_mocks = []

    def register(self, name, variants, group="", docs=""):
        self._stories.append(
            {"id": slug(name), "name": name, "group": group, "docs": docs, "variants": variants}
        )

    def mock(self, route, render, status=200):
        """Registry-level mock merged into every variant. A variant's own mock
        for the same route wins."""
        self._global_mocks.append(mock(route, render, status))
        return self

    def _mocks_for(self, v):
        return self._global_mocks + v["mocks"]

    def _find(self, sid, vname):
        for s in self._stories:
            if s["id"] == sid:
                for v in s["variants"]:
                    if v["name"] == vname:
                        return v
        return None

    @property
    def blueprint(self):
        bp = Blueprint("swapbook", __name__)
        bp.add_url_rule("/_swapbook/manifest.json", "manifest", self._manifest)
        bp.add_url_rule("/_swapbook/preview/<sid>/<vname>", "preview", self._preview)
        bp.add_url_rule("/_swapbook/mocks/<sid>/<vname>", "mocks", self._mocks)
        bp.add_url_rule(
            "/_swapbook/mock/<sid>/<vname>/<int:index>", "mock", self._mock,
            methods=["GET", "POST", "PUT", "DELETE", "PATCH"],
        )
        return bp

    def _manifest(self):
        def vmeta(v):
            m = {"name": v["name"], "controls": v["controls"], "docs": v["docs"]}
            if v.get("play"):
                m["play"] = v["play"]
            return m

        out = {
            "htmxSrc": self.htmx_src, "cssSrc": self.css_src, "jsSrc": self.js_src,
            "stories": [{
                "id": s["id"], "name": s["name"], "group": s["group"], "docs": s["docs"],
                "variants": [vmeta(v) for v in s["variants"]],
            } for s in self._stories],
        }
        if self.viewports:
            out["viewports"] = self.viewports
        return jsonify(out)

    def _preview(self, sid, vname):
        v = self._find(sid, vname)
        if not v:
            return Response(status=404)
        return Response(v["render"](_coerce(v["controls"], request.args.get)), mimetype="text/html")

    def _mocks(self, sid, vname):
        v = self._find(sid, vname)
        if not v:
            return Response(status=404)
        mks = self._mocks_for(v)
        return jsonify([{"verb": m["verb"], "path": m["path"], "index": i} for i, m in enumerate(mks)])

    def _mock(self, sid, vname, index):
        v = self._find(sid, vname)
        if not v:
            return Response(status=404)
        mks = self._mocks_for(v)
        if index < 0 or index >= len(mks):
            return Response(status=404)
        mk = mks[index]
        return Response(mk["render"]({}), status=mk.get("status", 200), mimetype="text/html")
